[#7] [Backend] As a user, I can receive and verify an verification e-mail

toby-thanathip opened the pull request:
Close #7

## What happened 👀

- Send verification e-mail after sign-up (with [letter_opener](https://github.com/ryanb/letter_opener) for development, otherwise with [mail-gun](https://github.com/mailgun/mailgun-ruby))
- Require that the e-mail has been verified before the user can sign in (with the help of `user.confirmed?`)
- Navigate to root when the confirmation link has been clicked

## Insight 📝

I had to add `include ActionController::Flash` in `ConfirmationsController` in order for the test to succeed 💭 

Seems like Devise is expecting this when the user is being redirected after confirmation 🤔 

**Note**: Since we don't have `staging.rb`, it will fallback to `production.rb`

## Proof Of Work 📹

**DEVELOPMENT (letter-opener):**

https://github.com/nimblehq/ic-rails-huey-toby/assets/18277915/98562eab-b988-4697-9c55-8a5577e0fe42

**STAGING/PRODUCTION (mail-gun):**

https://github.com/nimblehq/ic-rails-huey-toby/assets/18277915/79d49452-3422-4431-8a1e-206780eaa088

toby-thanathip committed:
[#7] Add confirmable for e-mail verification

hoangnguyen92dn commented:
@Tuubz This PR should belongs to #7 🙈

Tuubz commented:
@hoangnguyen92dn Hahah, sorryyy. Updated ✅

toby-thanathip committed:
[#7] Add tests for e-mail verification

Tuubz commented:
@hoangnguyen92dn @hoangmirs Resolved conflicts 🚀

hoangnguyen92dn reviewed (commented):
I am just curious that we don't need to setup email sender to send email to user's mailbox? 🤔

Tuubz commented:
> I am just curious that we don't need to setup email sender to send email to user's mailbox? 🤔

You mean sending a real e-mail (for staging/production), right? Good question, I didn't thank about 💭

Tuubz commented:
Sorry guys, I missed this requirement 🙈 I'll put this PR on draft for now and start working on it 🙏

toby-thanathip committed:
[#7] Send actual e-mail with Mailgun for staging/production

hoangnguyen92dn reviewed (approved):
@Tuubz From the PoW, the confirmation email should be redirected user to staging/production environment url instead of `localhost:3000`, so please don't forget to update the configs on heroku 😄

toby-thanathip commented:
@hoangnguyen92dn Thank you, I forgot to mention:

- I added the environment variables to staging/production 👍 
- For testing, I changed the values of `development.rb` to test it locally 🙏

olivierobert commented on Gemfile:
```diff
@@ -55,6 +55,7 @@ group :development, :test do
 
   # Utilities
   gem 'figaro' # Simple Rails app configuration
+  gem 'mailgun-ruby' # Contains methods for easily interacting with the Mailgun API
```
This should not be in the dev and test environments group only 🙈 This gem would not be installed in the prod environment. It should be moved in the block for `backend` (starting at line 4)

toby-thanathip committed:
[#7] Move maligun gem available for production environment

hoangmirs commented on app/controllers/api/v1/users/sessions_controller.rb:
```diff
@@ -32,6 +35,20 @@ def render_success(user)
+
+        def render_invalid_password_error
+          render_error(
+            detail: [I18n.t('devise.failure.invalid', authentication_keys: :email)],
```
Sorry, maybe I missed it, but why did we use the array here?

hoangmirs commented on spec/requests/sessions_controller_spec.rb:
```diff
@@ -55,6 +55,42 @@
+        }
+
+        expect(response).to have_http_status(:unauthorized)
+        expect(JSON.parse(response.body)['errors'][0]['detail']).to include(I18n.t('devise.failure.invalid', authentication_keys: :email))
```
The detail attribute should be a regular text following `JSON:API` spec:
> a human-readable explanation specific to this occurrence of the problem

And the test can be rewritten as:
```suggestion
        expect(JSON.parse(response.body)['errors'][0]['detail']).to eq(I18n.t('devise.failure.invalid', authentication_keys: :email))
```

toby-thanathip committed:
[#7] Improve tests
